### PR TITLE
Refactor C->C to C->R twice

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56964 Clean up naming and comments
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* **#55692 Refactor C->C to C->R twice**

### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output

Differential Revision: [D28004614](https://our.internmc.facebook.com/intern/diff/D28004614)

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -182,40 +182,25 @@ def compute(a, b):
...
-                # placeholder if grad_out is complex but output is not
-                jacobians_cols.append(torch.zeros_like(ds_dx))
-                continue
+            assert(not ds_dx.is_complex())
             # conjugate wirtinger derivative
             conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)
-            # wirtinger derivative
-            w_d = 0.5 * (ds_dx - ds_dy * 1j)
-            jacobians_cols.append(grad_out.conjugate() * conj_w_d + grad_out * w_d.conj())
+            jacobians_cols.append(2 * grad_out * conj_w_d)
```
Comment (albanD) [inline torch/autograd/gradcheck.py:199]: You can remove the 0.5 above and the 2 at this line :)

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -1033,19 +1004,16 @@ def gradcheck_helper(func, inputs, eps, atol, rtol, check_sparse_nnz, nondet_tol
     check_inputs(tupled_inputs, check_sparse_nnz)
     func_out = func(*tupled_inputs)
 
-    if fast_mode and has_complex_inputs_or_outputs(tupled_inputs, func_out):
-        raise NotImplementedError("Fast mode for gradcheck and gradgradcheck is currently only implemented"
-                                  " for R to R functions.")
+    # if fast_mode and has_complex_inputs_or_outputs(tupled_inputs, func_out):
```
Comment (albanD) [inline torch/autograd/gradcheck.py:1007]: Should this be deleted?

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -1033,19 +1004,16 @@ def gradcheck_helper(func, inputs, eps, atol, rtol, check_sparse_nnz, nondet_tol
...
     outputs = _differentiable_outputs(func_out)
     check_outputs(outputs)
-
-    if fast_mode:
-        fast_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
-                       atol, check_grad_dtypes, nondet_tol)
-    else:
-        slow_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
-                       atol, check_grad_dtypes, nondet_tol)
+    complex_indices = [i for i, o in enumerate(outputs) if o.is_complex()]
```
Comment (albanD) [inline torch/autograd/gradcheck.py:1013]: Do we actually have any example of a function that returns mixed real and complex outputs? Could we assume they are either all real or all complex?

**Review by albanD — COMMENTED** [review]
I like when a refactor removes a lot more code than it adds :D

File: torch/autograd/gradcheck.py [hunk 4]
```diff
@@ -182,40 +182,25 @@ def compute(a, b):
...
 def compute_numerical_jacobian_cols(jvp_fn, delta, input_is_complex, grad_out) -> List[torch.Tensor]:
     # Computing the jacobian only works for pure real or pure imaginary delta
-    # for details on the algorithm used here, refer:
+    # For details on the algorithm used here, refer:
     # Section 3.5.3 https://arxiv.org/pdf/1701.00392.pdf
     # s = fn(z) where z = x for real valued input
     # and z = x + yj for complex valued input
     jacobians_cols: List[torch.Tensor] = []
     ds_dx_tup = jvp_fn(delta)
+    assert(not isinstance(grad_out, complex))
```
Comment (albanD) [inline torch/autograd/gradcheck.py:191]: All the grad_out logic should be removed here right? This function is only ever called with grad_out=1 IIUC

File: torch/autograd/gradcheck.py (hunk 2 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:1007]: Whoops

File: torch/autograd/gradcheck.py (hunk 3 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:1047]: As discussed offline, keeping this since it is tested and doesn't add too much complexity

File: torch/autograd/gradcheck.py [hunk 5]
```diff
@@ -162,10 +162,13 @@ def get_numerical_jacobian(fn, inputs, target=None, eps=1e-3, grad_out=1.0):
                   "in a future version of PyTorch. If you have a specific use for "
                   "this or feature request for this to be a stable API, please file "
                   "us an issue at https://github.com/pytorch/pytorch/issues/new")
+    if grad_out != 1.0:
```
Comment (albanD) [inline torch/autograd/gradcheck.py:165]: nit: could you add a comment here saying that the parameter is only kept for BC reasons

File: torch/autograd/gradcheck.py [hunk 6]
```diff
@@ -424,15 +408,21 @@ def get_analytical_jacobian(inputs, output, nondet_tol=0.0, grad_out=1.0):
...
```
Comment (albanD) [inline torch/autograd/gradcheck.py:411]: Same comment about BC

File: torch/autograd/gradcheck.py [hunk 7]
```diff
@@ -961,8 +938,8 @@ def gradcheck(
     For complex functions, no notion of Jacobian exists. Gradcheck verifies if the numerical and
     analytical values of Wirtinger and Conjugate Wirtinger derivative are consistent. The gradient
     computation is done under the assumption that the overall function has a real valued output.
-    For functions with complex output, gradcheck compares the numerical and analytical gradients
-    for two values of :attr:`grad_output`: 1 and 1j. For more details, check out
+    For functions with complex output, gradcheck treats them as if they are two separate functions
```
Comment (albanD) [inline torch/autograd/gradcheck.py:941]: Maybe we can phrase this better?
gradcheck is applied to two real-valued functions corresponding to taking the real part of the complex outputs  for the first one and the imaginary part of the complex outputs for the second one.

File: torch/autograd/gradcheck.py [hunk 8]
```diff
@@ -725,47 +716,65 @@ def _differentiable_outputs(x):
     return tuple(o for o in _as_tuple(x) if o.requires_grad)
 
 
-def get_notallclose_msg(analytical, numerical, output_idx, input_idx, error_str='') -> str:
-    return error_str + 'Jacobian mismatch for output %d with respect to input %d,\n' \
+def get_notallclose_msg(analytical, numerical, output_idx, input_idx, complex_indices, inp_is_complex) -> str:
+    out_is_imag = complex_indices and output_idx in complex_indices
+    grad_output = "1j" if out_is_imag else "1"
+    prefix = "" if not inp_is_complex and not out_is_imag else \
+        f"Gradients failed to compare equal for grad output = {grad_output}"
```
Comment (albanD) [inline torch/autograd/gradcheck.py:723]: Well, we shouldn't talk about grad_output here anymore.
Maybe only "while considering the real part of the complex outputs only"

File: torch/autograd/gradcheck.py [hunk 9]
```diff
@@ -725,47 +716,65 @@ def _differentiable_outputs(x):
...
+
+        imag_func_out = imag_fn(*tupled_inputs)
+        imag_outputs = _differentiable_outputs(imag_func_out)
+        gradcheck_fn(imag_fn, imag_func_out, tupled_inputs, imag_outputs, eps,
+                     rtol, atol, check_grad_dtypes, nondet_tol, complex_indices)
+
+        real_func_out = real_fn(*tupled_inputs)
+        real_outputs = _differentiable_outputs(real_func_out)
+        gradcheck_fn(real_fn, real_func_out, tupled_inputs, real_outputs, eps,
+                     rtol, atol, check_grad_dtypes, nondet_tol)
```
Comment (albanD) [inline torch/autograd/gradcheck.py:757]: You're missing complex_indices no?

**Review by albanD — COMMENTED** [review]
LGTM

Could you add one test to make sure that the right error is raised?

File: torch/autograd/gradcheck.py (hunk 9 above)
Comment (soulitzer) [reply torch/autograd/gradcheck.py:757]: Good catch!

Comment (soulitzer) [reply torch/autograd/gradcheck.py:757]: This isn't being tested though, which is why we did not catch this. We'd need a function that fails while considering the real part of a complex output, but succeeds when considering only the imaginary part, or maybe there is some way to accomplish this with hooks.

File: test/test_autograd.py [hunk 10]
```diff
@@ -4246,10 +4246,13 @@ def fn(inputs):
         self.assertEqual(jacobian[0], 2 * torch.eye(4))
 
         with self.assertWarnsRegex(UserWarning, "get_numerical_jacobian was part of PyTorch's private API"):
-            jacobian = get_numerical_jacobian(fn, (a, b), eps=1e-6)
+            jacobian = get_numerical_jacobian(fn, (a, b), eps=1e-6, grad_out=1)
         self.assertEqual(jacobian[0], 2 * torch.eye(4))
         self.assertEqual(jacobian[1], 1 * torch.eye(4))
 
+        with self.assertRaisesRegex(ValueError, "Expected grad_out to be 1.0"):
```
Comment (soulitzer) [inline test/test_autograd.py:4253]: New test added to make sure right error is raised

File: torch/autograd/gradcheck.py (hunk 9 above)
Comment (albanD) [reply torch/autograd/gradcheck.py]: Yes I think you can add some hooks on the result and add noise to either the real or imaginary part of the gradients?

**Review by albanD — APPROVED** [review]
Thanks for the update! Looks good!

You will need to fix the lint though.
